### PR TITLE
fix: align json output with jest's

### PR DESCRIPTION
**Pull request by zxch3n**
In the current version(v0.7.11), the JSON reporter's output format deviates from jest's. See [Stackblitz example](https://stackblitz.com/edit/vitest-dev-vitest-gsjy9q?file=jest_output.json).

In PR https://github.com/vitest-dev/vitest/pull/489 json reporter is added, it implements the subset of jest's [AggregatedResult type](https://github.com/facebook/jest/blob/fe5e91c1b86323ab02b3de9d322b77e65a64e75a/packages/jest-test-result/src/types.ts#L58-L75). However, jest --json output is using [FormattedTestResults](https://github.com/facebook/jest/blob/fe5e91c1b86323ab02b3de9d322b77e65a64e75a/packages/jest-test-result/src/types.ts#L138-L154). 

And each object in `testResults` should represents a file rather than a test case.

### CODE REVIEW

**Comment by patak-dev** [comment]
Thanks for the PR @zxch3n! Things like status being `pass` instead of `passed` looks quite important to correct. Interesting that no one noticed this before. I wonder if tools consuming this format are able to accept both.

@cexbrayat would you help us review this PR and check that it won't be a breaking change for your use case?

File: packages/vitest/src/node/reporters/json.ts [hunk 1]
```diff
@@ -60,20 +84,74 @@ export class JsonReporter implements Reporter {
...
+      const file = test.file
+      if (file) {
+        if (!fileToTestCases.has(file.filepath))
+          fileToTestCases.set(file.filepath, { file, tests: [] })
+
+        fileToTestCases.get(file.filepath)!.tests.push(test)
+      }
+    }
+
+    for (const [filepath, { tests, file }] of fileToTestCases) {
```
Comment (patak-dev) [inline packages/vitest/src/node/reporters/json.ts:100]: I think that we could avoid calling `getTests` and then reconstruct the file-to-tests map if we iterate by file here. See getTests that will give you all the tests flatten for a file
https://github.com/vitest-dev/vitest/blob/a1922a07c54bf3f3b2cd9e2f23376ba254f97bab/packages/vitest/src/utils/tasks.ts#L4

```js
for (const file of files) {
  const tests = getTests(file)
```

File: packages/vitest/src/node/reporters/json.ts [hunk 2]
```diff
@@ -60,20 +84,74 @@ export class JsonReporter implements Reporter {
...
+      })
+
+      testResults.push({
+        assertionResults,
+        startTime,
+        endTime,
+        status: tests.every(t =>
+          t.result?.state === 'pass'
+           || t.result?.state === 'skip'
+           || t.result?.state === 'run'
```
Comment (patak-dev) [inline packages/vitest/src/node/reporters/json.ts:132]: Wouldn't be an internal error to still have a `run` state here?

File: packages/vitest/src/node/reporters/json.ts (hunk 1 above)
Comment (zxch3n) [reply packages/vitest/src/node/reporters/json.ts:100]: Yeah. You're right

File: packages/vitest/src/node/reporters/json.ts (hunk 2 above)
Comment (zxch3n) [reply packages/vitest/src/node/reporters/json.ts:132]: I'm not very familiar with the internal logic. Should I add a failure message in this case?

Comment (patak-dev) [reply packages/vitest/src/node/reporters/json.ts:132]: Maybe leave it as failed?

Comment (zxch3n) [reply packages/vitest/src/node/reporters/json.ts]: Should we throw an error here or append a failure message to the result? Or do anything to let the user know?

Comment (patak-dev) [reply packages/vitest/src/node/reporters/json.ts]: I think we could emit a warning and set it as failed in the output json.

File: packages/vitest/src/node/reporters/json.ts [hunk 3]
```diff
@@ -47,33 +71,80 @@ export class JsonReporter implements Reporter {
...
-    const tests = getTests(files)
-    const numTotalTests = tests.length
+    const fileToTestCases = new Map<File, Test[]>()
 
+    for (const file of files) {
+      const relatedTests = getTests([file])
+      fileToTestCases.set(file, relatedTests)
+    }
+
+    const tests = [...fileToTestCases.values()].flat()
```
Comment (patak-dev) [inline packages/vitest/src/node/reporters/json.ts:81]: I think there is no problem in calling `getTests` twice. This line could remain as `const tests = getTests(files)` and the `fileToTestCases` map isn't needed then. You can directly use `for (const file of files) {` in line 94.

File: packages/vitest/src/node/reporters/json.ts (hunk 2 above)
Comment (zxch3n) [reply packages/vitest/src/node/reporters/json.ts]: done

**Review by cexbrayat — APPROVED** [review]
LGTM, thanks @zxch3n 👍

**Comment by krzysztofradomski** [comment]
Hey all. Just like to add wanted to weight in from a  hmm regular user's perspective ;)
In my current project we've recently switched from jest to vitest, very happy with the change, but the we could not consume the json output in Storybook... So I wrote a simple Storybook addon, which I published to npm just yesterday. This addon of mine depends on the current structure of the output file, and now today I've found this thread haha!

**Comment by patak-dev** [comment]
> Hey all. Just like to add wanted to weight in from a hmm regular user's perspective ;) In my current project we've recently switched from jest to vitest, very happy with the change, but the we could not consume the json output in Storybook... So I wrote a simple Storybook addon, which I published to npm just yesterday. This addon of mine depends on the current structure of the output file, and now today I've found this thread haha!

😬 
Well, at least it should be only about removing the addon, no? Have you tried this PR? It would be good to get another confirmation that it is working as expected.

**Review by patak-dev — APPROVED** [review]

**Comment by krzysztofradomski** [comment]
Hmm I might give this branch a try later today... Maybe I should just
update my addon - I was so proud of it, so not gonna ditch it so fast lol

pon., 28 mar 2022, 13:45 użytkownik patak ***@***.***>
napisał:

**Pull request merged**